### PR TITLE
MCR-2337 allow custom user attributes in jwt token

### PR DESCRIPTION
**Pull request by golsch**
[Link to jira](https://mycore.atlassian.net/browse/MCR-2337).

### CODE REVIEW

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on fe89de63fc8abfb330d330c3eda70677e42f793e! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/6416-pull_request-1150

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on c49179fa3535aaaca36709460798b0442b9769ff! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/6419-pull_request-1150

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on c49179fa3535aaaca36709460798b0442b9769ff! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/6419-pull_request-1150

File: mycore-base/src/main/java/org/mycore/common/MCRUserInformation.java [hunk 1]
```diff
@@ -53,4 +53,10 @@
      */
     String getUserAttribute(String attribute);
 
+    /**
+     * Get all attribute names if they are provided by the underlying user system
+     * @return user attribute names as String array or null if no value is defined
+     */
+    String[] getUserAttributeNames(); //TODO maybe more generic
+
```
Comment (yagee-de) [inline mycore-base/src/main/java/org/mycore/common/MCRUserInformation.java]: Do not extend the interface. We do not ask for a list of attributes but ask for a specific attribute.

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRSessionFilter.java [hunk 2]
```diff
@@ -314,12 +315,15 @@ public boolean isUserInRole(String role) {
...
         public String getUserAttribute(String attribute) {
-            if (MCRUserInformation.ATT_REAL_NAME.equals(attribute)) {
-                return jwt.getClaim("name").asString();
-            }
-            if (MCRUserInformation.ATT_EMAIL.equals(attribute)) {
-                return jwt.getClaim("email").asString();
+            if (!jwt.getClaim(attribute).isNull()) {
+                return jwt.getClaim(attribute).asString();
+            } else {
+                return null;
```
Comment (yagee-de) [inline mycore-restapi/src/main/java/org/mycore/restapi/MCRSessionFilter.java]: leave mapping of standard attribute names to defined claims

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRSessionFilter.java [hunk 3]
```diff
@@ -314,12 +315,15 @@ public boolean isUserInRole(String role) {
 
         @Override
         public String getUserAttribute(String attribute) {
-            if (MCRUserInformation.ATT_REAL_NAME.equals(attribute)) {
-                return jwt.getClaim("name").asString();
-            }
-            if (MCRUserInformation.ATT_EMAIL.equals(attribute)) {
-                return jwt.getClaim("email").asString();
+            if (!jwt.getClaim(attribute).isNull()) {
+                return jwt.getClaim(attribute).asString();
```
Comment (yagee-de) [inline mycore-restapi/src/main/java/org/mycore/restapi/MCRSessionFilter.java]: Not all attribute names are valid claim names. We should not mix this and be more careful: https://auth0.com/docs/tokens/json-web-tokens/json-web-token-claims

**Review by yagee-de — CHANGES_REQUESTED** [review]

File: mycore-base/src/main/java/org/mycore/common/MCRUserInformation.java (hunk 1 above)
Comment (golsch) [reply mycore-base/src/main/java/org/mycore/common/MCRUserInformation.java]: Which attributes there are depends on the application. I don't know that explicitly at this point, that's why I chose the approach. Alternatively, prefixes for attributes could be explicitly defined in the configuration and queried. To solve this cleanly would actually also implement an adjustment to the interface or something hackier may query an attribute based on the configuration with a prefix for example with `accesskey*`.

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRSessionFilter.java (hunk 3 above)
Comment (golsch) [reply mycore-restapi/src/main/java/org/mycore/restapi/MCRSessionFilter.java]: Oh, right. I will additionally extend the claim names with a prefix.

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on 810a85186c1e3fb6fa5a43f05ec6ec9e628231e4! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/6553-pull_request-1150

**Pull request closed without merging**